### DISCUSSION TITLE
Dividing images to train,test and val

### DISCUSSION THREAD

**Discussion by SaadAliHafiz** [lukas-blecher/LaTeX-OCR#92] [Q&A] [answered]
![Capture](https://user-images.githubusercontent.com/59177839/152155516-5e6020aa-4b85-4e6e-b452-a54462f0d787.PNG)

greetings Lukas,
 `!gdown -O dataset/data/crohme.zip --id 13vjxGYrFCuYnwgDIUqkxsNGKk__D_sOM
!gdown -O dataset/data/pdf.zip --id 176PKaCUDWmTJdQwc-OfkO0y8t4gLsIvQ
!gdown -O dataset/data/pdfmath.txt --id 1QUjX6PFWPa-HBWdcY-7bA5TRVUnbyS1D
os.chdir('dataset/data')
!unzip -q crohme.zip 
!unzip -q pdf.zip 
os.chdir('images')
!mkdir ../valimages
!ls | shuf -n 1000 | xargs -i mv {} ../valimages
os.chdir('../../..')`
 by running this command, test, train and val directories are made, but its not of handwritten images.

**Comment by lukas-blecher** [comment, marked as answer]
Because the handwritten images are located in `images/` and `valimages/`, but you're looking at `trian/` which are the regular images.
I clearly stated in the notebook that the model will be trained on the mixed domain.

**Comment by SaadAliHafiz** [comment]
Got it thanks,
Lukas, I tried training only on handwritten images only I didn't include pdf equations but It's still giving me correct predictions on pdf equations while not recognizing handwritten equations why?
is it coz of we downloaded pretrained model before training ? by this command
`!curl -L -o checkpoints/weights.pth https://github.com/lukas-blecher/LaTeX-OCR/releases/download/v0.0.1/weights.pth`

**Reply by lukas-blecher** [reply]
What are your validation metrics for each? I can't imagine that the model is not learning anything. For how long did you train?
Starting with the pretrained model is supposed to cut some initial training and I would not recommend cutting this line.

**Reply by SaadAliHafiz** [reply]
![image](https://user-images.githubusercontent.com/59177839/152192830-debdaee5-a868-42e5-ba6b-f6603494488f.png)
and bleu score was 0.001

**Reply by SaadAliHafiz** [reply]
ive trained for 5 epochs

**Reply by lukas-blecher** [reply]
5 epochs might be too little, but BLEU score of 0.001 is not possible. Are you sure you evaluated the correct model?
I've just quickly finetuned the model for 10 epochs on CROHME and the BLEU score is 0.68 after 5 epochs and 0.73 after 10 epochs with 0.3 edit distance.
Here the command
```
python eval.py -c checkpoints/mixed/mixed_e05.pth --config settings/colab.yaml -d dataset/data/val.pkl
```

**Reply by SaadAliHafiz** [reply]
okie I'll try this

**Reply by SaadAliHafiz** [reply]
> 5 epochs might be too little, but BLEU score of 0.001 is not possible. Are you sure you evaluated the correct model? I've just quickly finetuned the model for 10 epochs on CROHME and the BLEU score is 0.68 after 5 epochs and 0.73 after 10 epochs with 0.3 edit distance. Here the command
> 
> ```
> python eval.py -c checkpoints/mixed/mixed_e05.pth --config settings/colab.yaml -d dataset/data/val.pkl
> ```

one more thing Lukas have u tried to generate the latex code against handwritten equation after the evaluation?

**Reply by lukas-blecher** [reply]
That's what the evaluation does. I don't have more data than the validation set.

**Reply by SaadAliHafiz** [reply]
> That's what the evaluation does. I don't have more data than the validation set.

aight lukas!

**Reply by SaadAliHafiz** [reply]
![image](https://user-images.githubusercontent.com/59177839/152217846-667ef410-3121-433d-aa83-af22fd3f256c.png)
So I did what u suggested me to do and got 0.7 bleu score but when I'm trying to predict the latex by this code 
`imgs = upload_files()
predictions = []
for name, f in imgs:
    img = Image.open(f)
    math = pix2tex.call_model(*args, img)
    print(math)
    predictions.append('\\mathrm{%s} & \\displaystyle{%s}'%(name, math))
Math(table%'\\\\'.join(predictions))`
But I'm getting wrong results even on the pictures I've trained on it

I've done this
![image](https://user-images.githubusercontent.com/59177839/152218331-e2dc1c52-8c2d-4188-87e8-a8bf774a2c6d.png)

**Reply by lukas-blecher** [reply]
Because you don't use the above trained model but the pretrained from GitHub? You do understand that everything but the structure of the model is stored in the checkpoint files, right? And if you don't specify said file it will most likely be the pretrained model from GitHub.

**Reply by SaadAliHafiz** [reply]
> Because you don't use the above trained model but the pretrained from GitHub? You do understand that everything but the structure of the model is stored in the checkpoint files, right? And if you don't specify said file it will most likely be the pretrained model from GitHub.

![image](https://user-images.githubusercontent.com/59177839/152225210-cea84170-99a6-40d5-bdcb-c62942a71190.png)
So how do I use my trained model? Ik my model path is the latest checkpoint folder all I want is to make the model recognize at least on the equation it'd trained. I don't want it to predict handwritten equations from other sources(just want prediction from crohme dataset equations).
can you specify the commands which could help me

**Reply by lukas-blecher** [reply]
If you are using two different colab sessions the command you used is invalid. You need to download the checkpoint file and save it onto your local hard drive. I'd recommend you stay on the local machine but if it is necessary you can also upload the file to the other colab session.
Now you can call either `pix2tex.py` or `gui.py` with an extra `checkpoint` argument:
```
python pix2tex.py --checkpoint path/to/checkpoint.pth
```
If you need to use Colab upload the checkpoint file and save it as `weights.pth` into the checkpoints directory.

**Reply by SaadAliHafiz** [reply]
> If you are using two different colab sessions the command you used is invalid. You need to download the checkpoint file and save it onto your local hard drive. I'd recommend you stay on the local machine but if it is necessary you can also upload the file to the other colab session. Now you can call either `pix2tex.py` or `gui.py` with an extra `checkpoint` argument:
> 
> ```
> python pix2tex.py --checkpoint path/to/checkpoint.pth
> ```
> 
> If you need to use Colab upload the checkpoint file and save it as `weights.pth` into the checkpoints directory.

![image](https://user-images.githubusercontent.com/59177839/152229993-f0135ffc-9c0f-4988-a037-cc49f42ee2bc.png)
like this?

**Reply by lukas-blecher** [reply]
Not exactly what I meant, but sure, if it works. Is `weights.pth` the new checkpoint file? I see the directory mixed, so probably better is:
```
python pix2tex.py --checkpoint checkpoints/mixed/mixed_e05.pth
```

**Reply by SaadAliHafiz** [reply]
> Not exactly what I meant, but sure, if it works. Is `weights.pth` the new checkpoint file? I see the directory mixed, so probably better is:
> 
> ```
> python pix2tex.py --checkpoint checkpoints/mixed/mixed_e05.pth
> ```

what does exactly checkpoint file mean can you briefly explain?

**Reply by SaadAliHafiz** [reply]
can you gimme the picture of handwritten generated latex ?

**Reply by lukas-blecher** [reply]
I mean the .pth file you trained, eg `mixed_e05.pth` or any other weights file. I meant you can also rename the file to weights.pth in the checkpoints dir.

**Reply by SaadAliHafiz** [reply]
I did, in my case, it was `mixed_e14.pth` so I use this

`!python pix2tex.py --checkpoint checkpoints/mixed/mixed_e14.pth`

**Reply by lukas-blecher** [reply]
ok perfect. Well did it work?

**Reply by SaadAliHafiz** [reply]
no :(
![image](https://user-images.githubusercontent.com/59177839/152232653-eac66fd8-24c0-4404-bfb8-9988f88191a1.png)

**Reply by lukas-blecher** [reply]
but BLEU is still good? What if you compute it for the training set as well? These should be exactly the same images.

**Reply by SaadAliHafiz** [reply]
Yep BLEU is still good, Okay I'll compute it on train.pkl

**Comment by SaadAliHafiz** [comment]
![image](https://user-images.githubusercontent.com/59177839/152239717-a91cc979-69ed-49e3-92ad-14d7840158f7.png)
both are same

**Reply by lukas-blecher** [reply]
That's very good. Don't know why it won't work for pix2tex though. Try adding the config file:
```
!python pix2tex.py --checkpoint checkpoints/mixed/mixed_e14.pth --config settings/colab.yaml
```

**Reply by SaadAliHafiz** [reply]
![Capture1](https://user-images.githubusercontent.com/59177839/152243070-b6b7390f-3978-4135-91ca-135a3621d529.PNG)
it finally worked thanks, Lukas!
You're a life saver

**Reply by lukas-blecher** [reply]
Nice :D Make sure to check the valimages too

**Reply by SaadAliHafiz** [reply]
I did and it worked on that as well

**Reply by SaadAliHafiz** [reply]
I've even tried to check on other handwritten images from google and it worked on those as well.
